internal/k8s/client.go

```diff
@@ -0,0 +1,166 @@
+package k8s
+
+import (
+	"errors"
+	"fmt"
+	"os"
+
+	"github.com/Shopify/kubeaudit/k8stypes"
+	log "github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/version"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+)
+
+// ErrNoReadableKubeConfig represents any error that prevents the client from opening a kubeconfig file.
+var ErrNoReadableKubeConfig = errors.New("unable to open kubeconfig file")
+
+var DefaultClient = K8sClient{}
+
+// Client abstracts the API to allow testing.
+type Client interface {
+	InClusterConfig() (*rest.Config, error)
+}
+
+// K8sClient wraps kubernetes client-go so it can be mocked.
+type K8sClient struct{}
+
+// InClusterConfig wraps the client-go method with the same name.
+func (kc K8sClient) InClusterConfig() (*rest.Config, error) {
+	return rest.InClusterConfig()
+}
+
+func NewKubeClientLocal(configPath string) (*kubernetes.Clientset, error) {
+	if _, err := os.Stat(configPath); err != nil {
+		return nil, ErrNoReadableKubeConfig
+	}
+	config, err := clientcmd.BuildConfigFromFlags("", configPath)
+	if err != nil {
+		return nil, err
+	}
+	kube, err := kubernetes.NewForConfig(config)
+	return kube, err
+}
+
+func NewKubeClientCluster(client Client) (*kubernetes.Clientset, error) {
+	config, err := client.InClusterConfig()
+	if err != nil {
+		return nil, err
+	}
+	log.Info("Running inside cluster, using the cluster config")
+	kube, err := kubernetes.NewForConfig(config)
+	return kube, err
+}
+
+func IsRunningInCluster(client Client) bool {
+	config, err := client.InClusterConfig()
+	fmt.Println("Config ", config)
+	return err == nil
+}
+
+func GetAllResources(clientset kubernetes.Interface, namespace string) []k8stypes.Resource {
+	var resources []k8stypes.Resource
+
+	for _, resource := range GetDaemonSets(clientset, namespace).Items {
```

Comment:
Could you make a type out of the `Get*` functions that have the same signature? 

```golang
type resourceGetter func(clientset kubernetes.Interface, namespace string) *k8stypes.DaemonSetListV1
```

And then replace this with:

```golang
for _, resourceGetter := range []resourceGetter{
    GetDaemonSets,
    GetDeployments,
} {
       for _, resource := range resourceGetter(clients, namespace).Items {
            resources = append(resources, resource.DeepCopyObject())
       }
}
```

Reply:
I think when I tried to do this I got stuck on the fact that `Items` is a struct field and not a function call so what would the return value of the shared function signature be?

Reply:
The returned value would be whatever structure contains `Items`!

Reply:
Hmm but it's a different structure for each resource type?
DaemonSetList https://godoc.org/k8s.io/api/apps/v1#DaemonSetList
Deployment https://godoc.org/k8s.io/api/apps/v1#DeploymentList

Reply:
Again, open an issue do it in a different PR?

Reply:
Sorry, I thought I responded! I didn't realize they were different structures, ignore me! I'm not even certain this warrants an issue, I thought it was a quick optimization but looking at those structures it does not look it.

Reply:
Sorry, I thought I responded! I didn't realize they were different structures, ignore me! I'm not even certain this warrants an issue, I thought it was a quick optimization but looking at those structures it does not look it.

Edit: attached this comment to the wrong thread...